File: Pancake_Sort/Pancake_Sort.rb
# Function to find the maximum index
def max_index(arr , n)
    mi = 0
    for i in 0..n-1
        if arr[i]>arr[mi]
            mi = i
        end
    end
    return mi+1
end

#Function to flip the give array
def flip(arr , m)
    start = 0
    last = m-1
    while start<last
        a = arr[start]
        arr[start] = arr[last]
        arr[last] = a
        start = start+1
        last = last-1
    end
    return arr
end

# Function to sort the given input
def pancake_sort(arr , len)
    n = len
    m = len
    perm_arr = arr
    while m>1
        mi = max_index(perm_arr , m)
        if mi != m-1
            # Flips the array from index 0 till the max index
            temp_arr = flip(perm_arr , mi)
            # Flips the array from index 0 till the length of the array
            perm_arr = flip(temp_arr , m)
        end
        m = m-1
    end
    for i in 0..n-1
        print(perm_arr[i].to_s + " ")
    end
end

input = [2,7,8,1,4,1]
len = input.length()

# Driver program
pancake_sort(input , len)

=begin
input = [2,7,8,1,4,1]
output = 1 1 2 4 7 8
=end

File: Pancake_Sort/Pancake_Sort.dart
/**
 *   Pancake Sort
 * --------------------------
 * It is a sorting algorithm of O(n^2) algorithm which sorts like selection sort
 * by moving the minimum to the front by reversing the array till its index.
 */

// Importing required libraries
import 'dart:io';

// Utility function to find maximum of the array
int findMax(List array, int n){
    int max_value = array[0];
    int max_index = 0;
    for (int i = 0; i < n; i ++){
        if (max_value < array[i]){
            max_value = array[i];
            max_index = i;
        }
    }
    return max_index;
}

// Utility function to flip an array from an index
void flip(List array, int i){
    int start = 0, temp;
    while (start < i){
        temp = array[start];
        array[start] = array[i];
        array[i] = temp;
        start ++;
        i --;
    }
}

// Function to sort a list using pancake sort
void pancakeSort(List array, int n){
    for (int i = n; i > 1; i --){
        int max_index = findMax(array, i);
        if (max_index != i - 1){
            flip(array, max_index);
            flip(array, i - 1);
        }
    }
}
// Driver method of the program
void main(){
    print("Enter number of elements of the array:");
    var input = stdin.readLineSync();
    int n = int.parse(input);

    print("Enter array elements:");
    input = stdin.readLineSync();
    var lis = input.split(" ");
    List array = lis.map(int.parse).toList();

    pancakeSort(array, n);

    print(array.join(" "));
}

/**
 *   Sample Input and Output
 * ------------------------------------
 * Enter number of elements of the array:
 * 10
 * Enter array elements:
 * 10 9 8 7 6 5 4 3 2 1
 * 1 2 3 4 5 6 7 8 9 10
 */
